### PR TITLE
fix: save and restore pb state

### LINKED ISSUE #9270: help request: opentelemetry + grpc-transcode not working
### Description

I've been testing apisix and tried using otel plugin while transcoding to grpc and got an error. I did try using the otel plugin with http and it works fine, so I assume the problem is with the 2 plugins working together. This is the error i got:
```
 2023/04/08 05:21:57 [error] 56#56: *4242 lua entry thread aborted: runtime error: ...deps/share/lua/5.1/opentelemetry/trace/exporter/otlp.lua:104: bad argument #1 to 'encode' (type 'opentelemetry.proto.trace.v1.TracesData' does not exists)
 stack traceback:
 coroutine 0:
        [C]: in function 'encode'
        ...deps/share/lua/5.1/opentelemetry/trace/exporter/otlp.lua:104: in function 'export_spans'
        ...are/lua/5.1/opentelemetry/trace/batch_span_processor.lua:45: in function 'process_batches'
        ...are/lua/5.1/opentelemetry/trace/batch_span_processor.lua:77: in function <...are/lua/5.1/opentelemetry/trace/batch_span_processor.lua:57>, context: ngx.timer, client: 172.20.0.1, server: 0.0.0.0:9080
```

This is my config:
```
# config.yaml
plugins:
  - grpc-transcode
  - opentelemetry

  opentelemetry:
    resource:
      service.name: APISIX
      tenant.id: business_id
    collector:
      address: otel-collector:4318
      request_timeout: 3
      request_headers:
        foo: bar
    batch_span_processor:
      drop_on_queue_full: false
      max_queue_size: 6
      batch_timeout: 2
      inactive_timeout: 1
      max_export_batch_size: 2
```
```
# apisix.yaml
routes:
  - id: users
    methods: [GET]
    uri: /users
    plugins:
      grpc-transcode:
        proto_id: "1"
        service: users.UsersService
        method: FindAll
      opentelemetry:
        sampler:
          name: always_on
    upstream:
      scheme: grpc
      type: roundrobin
      nodes:
        "users:5000": 1
```
The transcoding part works fine, I'm able to get the expected response from the grpc service, just missing the otel trace.
Am I maybe using the plugins incorrectly?

### Environment

- APISIX version (run `apisix version`): 3.2.0 docker image
- Operating system (run `uname -a`): Ubuntu WSL2

### PR DESCRIPTION
**Pull request by kingluo**
### Description

Fixes #9270

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by leslie-tsang — DISMISSED** [review]

File: t/plugin/opentelemetry-bugfix-pb-state.t
Comment (monkeyDluffy6017) [inline t/plugin/opentelemetry-bugfix-pb-state.t:1]: It's a opentelemetry or grpc-transcode problem, add the test cases to opentelemetry2.t or grpc-transcode3.t, no need to create a new test file.

File: apisix/plugins/grpc-transcode/response.lua [hunk 1]
```diff
@@ -99,7 +99,10 @@ return function(ctx, proto, service, method, pb_option, show_status_in_body, sta
 
     -- handle error response after the last response chunk
     if ngx.status >= 300 and show_status_in_body then
-        return handle_error_response(status_detail_type)
+        local pb_old_state = pb.state(proto.pb_state)
+        local ret = handle_error_response(status_detail_type)
+        pb.state(pb_old_state)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/grpc-transcode/response.lua:104]: Why not do this in `handle_error_response` function, `around pd.decode` like other place.

File: t/plugin/opentelemetry-bugfix-pb-state.t
Comment (kingluo) [reply t/plugin/opentelemetry-bugfix-pb-state.t:1]: No, it doesn't belong to any plugin, this is a common pb status issue with any plugin. It's best to use a stand-alone test file to highlight this issue.

**Review by monkeyDluffy6017 — APPROVED** [review]

File: t/plugin/opentelemetry-bugfix-pb-state.t [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+add_block_preprocessor(sub {
+    my ($block) = @_;
+
+    if (!$block->extra_yaml_config) {
+        my $extra_yaml_config = <<_EOC_;
+plugins:
+    - example-plugin
+    - key-auth
+    - opentelemetry
+plugin_attr:
```
Comment (monkeyDluffy6017) [inline t/plugin/opentelemetry-bugfix-pb-state.t:29]: Don't you need the grpc-transcode plugin?

Comment (kingluo) [reply t/plugin/opentelemetry-bugfix-pb-state.t:29]: No.
https://github.com/apache/apisix/blob/71c6048e3e95c0c85936e828b87fa50c5ba87d1b/apisix/admin/proto.lua#L23
Mark my last words. It does not depend on a specific plugin. Instead, it's a general problem.

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**